Lib/test/test_zipimport.py

```diff
@@ -669,6 +669,26 @@ def testBytesPath(self):
 class CompressedZipImportTestCase(UncompressedZipImportTestCase):
     compression = ZIP_DEFLATED
 
+    @support.cpython_only
+    def test_issue31602(self):
+        # There shouldn't be an assertion failure in zipimporter.get_source()
+        # in case of a bad zlib.decompress().
+        import zlib
+        def bad_decompress(*args):
+            return None
+        z = ZipFile(TEMP_ZIP, 'w')
+        try:
+            name = 'bar.py'
+            data = b'spam'
+            z.writestr(name, data, ZIP_DEFLATED)
+            z.close()
+            zi = zipimport.zipimporter(TEMP_ZIP)
+            with support.swap_attr(zlib, 'decompress', bad_decompress):
+                self.assertRaises(TypeError, zi.get_source, 'bar')
+        finally:
+            z.close()
```

Comment:
You also called `z.close()` above, so I would use my `self.addCleanup()` suggestion below and rework this to drop the `try` block and use the context manager for zipfiles:
```python
name = 'bar.py'
data = b'spam'
with ZipFile(...) as zip_file:
    self.addCleanup(os.remove, TEMP_ZIP)
    zip_file.writestr(...)
zi = ...
```